### PR TITLE
Code style options cleanup

### PR DESCRIPTION
**Pull request by tmat**
Today we have two interfaces that represent values of a code style options: ICodeStyleOption (public) and ICodeStyleOption2 (internal).

ICodeStyleOption is only used for options we hand out via legacy public APIs. It is not available in code style layer.
ICodeStyleOption2 should be used throughout Roslyn. The internal values are translated to the public ones at the public entry points.

ICodeStyleOption2 implemented ICodeStyleOption. This change separates these two types entirely, removing the dependency, and cleans up related code.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Options/OptionSet.cs [hunk 1]
```diff
@@ -93,11 +93,11 @@ bool IOptionsReader.TryGetOption<T>(OptionKey2 optionKey, out T value)
     /// Checks if the value is an internal representation -- does not cover all cases, just code style options.
     /// </summary>
     internal static bool IsInternalOptionValue(object? value)
-        => value is not ICodeStyleOption codeStyle || ReferenceEquals(codeStyle, codeStyle.AsInternalCodeStyleOption());
+        => value is not ICodeStyleOption codeStyle || codeStyle.GetType() == typeof(CodeStyleOption2<>);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Options/OptionSet.cs]: This is never going to match (need to get the generic type definition on the left side)

File: src/Workspaces/Core/Portable/Options/OptionSet.cs [hunk 2]
```diff
@@ -93,11 +93,11 @@ bool IOptionsReader.TryGetOption<T>(OptionKey2 optionKey, out T value)
...
 
     /// <summary>
     /// Checks if the value is an public representation -- does not cover all cases, just code style options.
     /// </summary>
     internal static bool IsPublicOptionValue(object? value)
-        => value is not ICodeStyleOption codeStyle || ReferenceEquals(codeStyle, codeStyle.AsPublicCodeStyleOption());
+        => value is not ICodeStyleOption codeStyle || codeStyle.GetType() != typeof(CodeStyleOption2<>);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Options/OptionSet.cs]: Same issue here

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Options/OptionDefinition.cs [hunk 3]
```diff
@@ -89,7 +89,10 @@ public static bool IsSupportedOptionType(Type type)
            type == typeof(long?) ||
            type.IsEnum ||
            Nullable.GetUnderlyingType(type)?.IsEnum == true ||
+#if !CODE_STYLE
            typeof(ICodeStyleOption).IsAssignableFrom(type) ||
+#endif
+           typeof(ICodeStyleOption2).IsAssignableFrom(type) ||
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Options/OptionDefinition.cs:95]: maybe doc.  i really don't know how to judge what is going on here.

**Comment by CyrusNajmabadi** [comment]
> ICodeStyleOption is only used for options we hand out via legacy public APIs. It is not available in code style layer.
ICodeStyleOption2 should be used throughout Roslyn. The internal values are translated to the public ones at the public entry points.

If this isn't already stated on both types, please add.  This is def critical info for wrapping ones head around what is going on inside roslyn :D

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**